src/sql/deserialize.rs

```diff
@@ -0,0 +1,1333 @@
+
+            stmt.execute(named_params! {
+                ":msg_id": msg_id,
+                ":contact_id": contact_id,
+                ":reaction": reaction,
+            })?;
+        }
+
+        Ok(())
+    }
+
+    async fn deserialize_sending_domains(&mut self, tx: &mut Transaction<'_>) -> Result<()> {
+        let mut stmt = tx.prepare(
+            "
+INSERT INTO sending_domains
+        (domain,  dkim_works)
+VALUES (:domain, :dkim_works)",
+        )?;
+
+        self.expect_list().await?;
+
+        while self.expect_dictionary_opt().await? {
+            self.expect_key("domain").await?;
+            let domain = self.expect_string().await?;
+
+            self.expect_key("dkim_works").await?;
+            let dkim_works = self.expect_i64().await?;
+
+            self.expect_end().await?;
+
+            stmt.execute(named_params! {
+                ":domain": domain,
+                ":dkim_works": dkim_works,
+            })?;
+        }
+
+        Ok(())
+    }
+
+    async fn deserialize_tokens(&mut self, tx: &mut Transaction<'_>) -> Result<()> {
+        let mut stmt = tx.prepare(
+            "
+INSERT INTO tokens
+        (id,  namespc,  foreign_id,  token,  timestamp)
+VALUES (:id, :namespc, :foreign_id, :token, :timestamp)",
+        )?;
+
+        self.expect_list().await?;
+
+        while self.expect_dictionary_opt().await? {
+            self.expect_key("foreign_id").await?;
+            let foreign_id = self.expect_u32().await?;
+
+            self.expect_key("id").await?;
+            let id = self.expect_i64().await?;
+
+            self.expect_key("namespace").await?;
+            let namespace = self.expect_u32().await?;
+
+            self.expect_key("timestamp").await?;
+            let timestamp = self.expect_i64().await?;
+
+            self.expect_key("token").await?;
+            let token = self.expect_string().await?;
+
+            self.expect_end().await?;
+
+            stmt.execute(named_params! {
+                ":id": id,
+                ":namespc": namespace,
+                ":foreign_id": foreign_id,
+                ":token": token,
+                ":timestamp": timestamp,
+            })?;
+        }
+
+        Ok(())
+    }
+
+    async fn skip_until_end(&mut self) -> Result<()> {
+        let mut level: usize = 0;
+        loop {
+            let token = self.expect_token().await?;
+            match token {
+                BencodeToken::End => {
+                    if level == 0 {
+                        return Ok(());
+                    } else {
+                        level -= 1;
+                    }
+                }
+                BencodeToken::ByteString(_) | BencodeToken::Integer(_) => {}
+                BencodeToken::List | BencodeToken::Dictionary => level += 1,
+            }
+        }
+    }
+
+    async fn deserialize(mut self, mut tx: Transaction<'_>) -> Result<()> {
+        self.expect_dictionary().await?;
+
+        self.expect_key("_config").await?;
```

Comment:
Since you're holding the `Transaction` over an `await` point, the future returned by the function will not be `Send`.

JsonRPC requires all futures to be `Send`, so this:

```diff
diff --git a/deltachat-jsonrpc/src/api/mod.rs b/deltachat-jsonrpc/src/api/mod.rs
index b71718299..88f19e504 100644
--- a/deltachat-jsonrpc/src/api/mod.rs
+++ b/deltachat-jsonrpc/src/api/mod.rs
@@ -146,6 +146,16 @@ async fn inner_get_backup_qr(&self, account_id: u32) -> Result<Qr> {
 
 #[rpc(all_positional, ts_outdir = "typescript/generated")]
 impl CommandApi {
+    async fn t(&self) {
+        let mut w = Vec::new();
+        self.get_context(0)
+            .await
+            .unwrap()
+            .sql
+            .serialize(&mut w)
+            .await;
+    }
+
     /// Test function.
     async fn sleep(&self, delay: f64) {
         tokio::time::sleep(std::time::Duration::from_secs_f64(delay)).await
diff --git a/src/context.rs b/src/context.rs
index 627e63c78..b2bf48129 100644
--- a/src/context.rs
+++ b/src/context.rs
@@ -188,7 +188,7 @@ fn deref(&self) -> &Self::Target {
 pub struct InnerContext {
     /// Blob directory path
     pub(crate) blobdir: PathBuf,
-    pub(crate) sql: Sql,
+    pub sql: Sql,
     pub(crate) smeared_timestamp: SmearedTimestamp,
     /// The global "ongoing" process state.
     ///
```
will give you a vey long and unreadable error.

https://docs.rs/tokio/latest/tokio/task/struct.LocalSet.html#use-with-run_until looks like it might make it possible to run a non-send function from JsonRPC, but I didn't research further for now. Maybe you know more about this, anyway.